File: src/shared/http_readonly.py
import asyncio
import hashlib
import json
import math
import random
import time
import uuid
from collections.abc import Mapping, Sequence
from typing import cast
from urllib.parse import unquote, urlsplit
from weakref import WeakKeyDictionary

import httpx

from src.shared.runtime import logger

type JsonScalar = str | int | float | bool | None
type JsonValue = JsonScalar | list[JsonValue] | dict[str, JsonValue]
type QueryParams = Mapping[str, JsonScalar | Sequence[JsonScalar]]

_ALLOWED_HOSTNAMES = frozenset({"dadosabertos.compras.gov.br", "pncp.gov.br"})
_RETRYABLE_STATUSES = frozenset({429, 502, 503, 504})
_RETRYABLE_KINDS = frozenset({"UPSTREAM_TIMEOUT", "UPSTREAM_UNAVAILABLE"})
_RETRY_BASE_SECONDS = 0.2
_RETRY_JITTER_RATIO = 0.1
_MAX_ERROR_BODY_CHARS = 4096


class _RequestLimiter:
    def __init__(
        self,
        max_concurrency: int | None,
        requests_per_second: float | None,
    ) -> None:
        self.semaphore = (
            asyncio.Semaphore(max_concurrency) if max_concurrency is not None else None
        )
        self.rate_lock = asyncio.Lock() if requests_per_second is not None else None
        self.request_interval = (
            1 / float(requests_per_second) if requests_per_second is not None else None
        )
        self.next_request_at = 0.0


_LIMITERS: WeakKeyDictionary[
    asyncio.AbstractEventLoop, dict[str, _RequestLimiter]
] = WeakKeyDictionary()


class UpstreamError(RuntimeError):
    def __init__(
        self,
        kind: str,
        status: int | None,
        message: str,
        provider: str | None = None,
        endpoint: str | None = None,
        upstream_message: str | None = None,
    ) -> None:
        super().__init__(f"{kind}: {message}")
        self.kind = kind
        self.status = status
        self.message = message
        self.provider = provider
        self.endpoint = endpoint
        self.upstream_message = (message if upstream_message is None else upstream_message)[
            :_MAX_ERROR_BODY_CHARS
        ]
        self.retryable = status in _RETRYABLE_STATUSES or kind in _RETRYABLE_KINDS


class ReadOnlyHttpClient:
    def __init__(
        self,
        base_url: str,
        max_retries: int = 3,
        timeout: float = 30,
        max_document_bytes: object = 25_000_000,
        max_concurrency: int | None = None,
        requests_per_second: float | None = None,
        provider: str | None = None,
    ) -> None:
        if max_retries < 0:
            raise ValueError("max_retries must be non-negative")
        if isinstance(max_document_bytes, bool) or not isinstance(max_document_bytes, int):
            raise ValueError("max_document_bytes must be a positive finite integer")
        if max_document_bytes <= 0:
            raise ValueError("max_document_bytes must be a positive finite integer")
        if (
            not isinstance(max_concurrency, int)
            or isinstance(max_concurrency, bool)
            or max_concurrency <= 0
        ) and max_concurrency is not None:
            raise ValueError("max_concurrency must be positive")
        if (
            not isinstance(requests_per_second, (int, float))
            or isinstance(requests_per_second, bool)
            or not math.isfinite(float(requests_per_second))
            or requests_per_second <= 0
        ) and requests_per_second is not None:
            raise ValueError("requests_per_second must be positive")

        parsed_base = urlsplit(base_url)
        try:
            port = parsed_base.port
        except ValueError as error:
            raise ValueError("base_url must use a supported HTTPS port") from error
        hostname = parsed_base.hostname.lower() if parsed_base.hostname is not None else None
        if (
            parsed_base.scheme.lower() != "https"
            or hostname not in _ALLOWED_HOSTNAMES
            or port not in {None, 443}
            or parsed_base.username is not None
            or parsed_base.password is not None
            or parsed_base.query
            or parsed_base.fragment
        ):
            raise ValueError("base_url must use an allowlisted HTTPS host and supported port")

        self._max_retries = max_retries
        self._max_document_bytes = max_document_bytes
        self._max_concurrency = max_concurrency
        self._requests_per_second = requests_per_second
        self._origin = f"{parsed_base.scheme.lower()}://{hostname}"
        self._provider = provider or ("pncp" if hostname == "pncp.gov.br" else "compras")
        self._client = httpx.AsyncClient(
            base_url=f"{base_url.rstrip('/')}/",
            timeout=timeout,
        )

    async def __aenter__(self) -> "ReadOnlyHttpClient":
        return self

    async def __aexit__(self, *_: object) -> None:
        await self._client.aclose()

    async def get(self, path: str, params: QueryParams | None = None) -> JsonValue | str:
        request_path = self._relative_path(path)
        endpoint = f"/{request_path}"
        request_id = uuid.uuid4().hex

        for attempt in range(self._max_retries + 1):
            retry = False
            status_code: int | None = None
            started = time.monotonic()
            limiter = self._get_limiter()
            if limiter is not None and limiter.semaphore is not None:
                await limiter.semaphore.acquire()
            try:
                await self._wait_for_rate_limit(limiter)
                started = time.monotonic()
                async with self._client.stream("GET", request_path, params=params) as response:
                    status_code = response.status_code
                    body = await self._read_body(response, endpoint)
                    if response.status_code in _RETRYABLE_STATUSES:
                        if attempt < self._max_retries:
                            retry = True
                        else:
                            raise UpstreamError(
                                self._error_kind(response.status_code),
                                response.status_code,
                                self._error_body_text(response, body),
                                provider=self._provider,
                                endpoint=endpoint,
                            )
                    elif response.status_code == 404:
                        raise UpstreamError(
                            "NOT_FOUND",
                            404,
                            self._error_body_text(response, body),
                            provider=self._provider,
                            endpoint=endpoint,
                        )
                    elif response.status_code >= 300:
                        raise UpstreamError(
                            self._error_kind(response.status_code),
                            response.status_code,
                            self._error_body_text(response, body),
                            provider=self._provider,
                            endpoint=endpoint,
                        )

                    if not retry:
                        if self._is_json(response):
                            try:
                                return cast(JsonValue, json.loads(body))
                            except ValueError as error:
                                raise UpstreamError(
                                    "UPSTREAM_SCHEMA_CHANGED",
                                    response.status_code,
                                    str(error),
                                    provider=self._provider,
                                    endpoint=endpoint,
                                ) from error
                        return self._body_text(response, body)
            except httpx.DecodingError as error:
                raise UpstreamError(
                    "UPSTREAM_SCHEMA_CHANGED",
                    status_code,
                    str(error),
                    provider=self._provider,
                    endpoint=endpoint,
                ) from error
            except (httpx.TimeoutException, TimeoutError) as error:
                if attempt < self._max_retries:
                    retry = True
                else:
                    raise UpstreamError(
                        "UPSTREAM_TIMEOUT",
                        None,
                        str(error),
                        provider=self._provider,
                        endpoint=endpoint,
                    ) from error
            except (httpx.NetworkError, OSError) as error:
                if attempt < self._max_retries:
                    retry = True
                else:
                    raise UpstreamError(
                        "UPSTREAM_UNAVAILABLE",
                        None,
                        str(error),
                        provider=self._provider,
                        endpoint=endpoint,
                    ) from error
            except httpx.TransportError as error:
                if attempt < self._max_retries:
                    retry = True
                else:
                    raise UpstreamError(
                        "UPSTREAM_UNAVAILABLE",
                        None,
                        str(error),
                        provider=self._provider,
                        endpoint=endpoint,
                    ) from error
            finally:
                if limiter is not None and limiter.semaphore is not None:
                    limiter.semaphore.release()
                logger.info(
                    "upstream request",
                    extra={
                        "request_id": request_id,
                        "provider": self._provider,
                        "endpoint": hashlib.sha256(endpoint.encode("utf-8")).hexdigest(),
                        "duration_ms": round((time.monotonic() - started) * 1000, 3),
                        "status_code": status_code,
                        "retry_count": attempt,
                        "cache_hit": False,
                    },
                )
            if retry:
                await self._sleep_before_retry(attempt)
                continue

        raise AssertionError("unreachable")

    def _get_limiter(self) -> _RequestLimiter | None:
        loop = asyncio.get_running_loop()
        limiters = _LIMITERS.setdefault(loop, {})
        limiter = limiters.get(self._origin)
        if limiter is None and (
            self._max_concurrency is not None or self._requests_per_second is not None
        ):
            limiter = _RequestLimiter(self._max_concurrency, self._requests_per_second)
            limiters[self._origin] = limiter
        return limiter

    async def _wait_for_rate_limit(self, limiter: _RequestLimiter | None) -> None:
        if limiter is None or limiter.rate_lock is None or limiter.request_interval is None:
            return
        async with limiter.rate_lock:
            now = time.monotonic()
            if limiter.next_request_at > now:
                await asyncio.sleep(limiter.next_request_at - now)
            limiter.next_request_at = time.monotonic() + limiter.request_interval

    @staticmethod
    def _relative_path(path: str) -> str:
        decoded = path
        while (next_decoded := unquote(decoded)) != decoded:
            decoded = next_decoded
        parsed = urlsplit(decoded)
        decoded_path = parsed.path
        if (
            not path.startswith("/")
            or parsed.scheme
            or parsed.netloc
            or decoded.startswith("//")
            or parsed.query
            or parsed.fragment
            or "\\" in decoded
        ):
            raise ValueError("path must be relative to the fixed upstream origin")
        if ".." in decoded_path.replace("\\", "/").split("/"):
            raise ValueError("path must be relative to the fixed upstream origin")
        return path.lstrip("/")

    @staticmethod
    def _is_json(response: httpx.Response) -> bool:
        content_type = response.headers.get("content-type", "").split(";", 1)[0].strip().lower()
        return content_type == "application/json" or content_type.endswith("+json")

    @staticmethod
    def _error_kind(status: int) -> str:
        if status == 429:
            return "UPSTREAM_RATE_LIMIT"
        if status >= 500:
            return "UPSTREAM_UNAVAILABLE"
        return "UPSTREAM_BAD_REQUEST"

    @staticmethod
    async def _sleep_before_retry(attempt: int) -> None:
        delay = _RETRY_BASE_SECONDS * (2**attempt)
        await asyncio.sleep(delay + random.uniform(0, delay * _RETRY_JITTER_RATIO))

    async def _read_body(self, response: httpx.Response, endpoint: str) -> bytes:
        content_length = response.headers.get("content-length")
        if content_length is not None:
            try:
                declared_length = int(content_length)
            except ValueError:
                declared_length = None
            if declared_length is not None and declared_length > self._max_document_bytes:
                raise UpstreamError(
                    "DOCUMENT_TOO_LARGE",
                    response.status_code,
                    f"response exceeds {self._max_document_bytes} bytes",
                    provider=self._provider,
                    endpoint=endpoint,
                )

        body = bytearray()
        async for chunk in response.aiter_bytes():
            if len(body) + len(chunk) > self._max_document_bytes:
                raise UpstreamError(
                    "DOCUMENT_TOO_LARGE",
                    response.status_code,
                    f"response exceeds {self._max_document_bytes} bytes",
                    provider=self._provider,
                    endpoint=endpoint,
                )
            body.extend(chunk)
        return bytes(body)

    @staticmethod
    def _body_text(response: httpx.Response, body: bytes) -> str:
        return body.decode(response.encoding or "utf-8", errors="replace")

    @staticmethod
    def _error_body_text(response: httpx.Response, body: bytes) -> str:
        return ReadOnlyHttpClient._body_text(response, body)[:_MAX_ERROR_BODY_CHARS]

File: src/shared/http_readonly.ts
import { randomBytes, createHash } from "node:crypto";

import { logger } from "./runtime.js";

export type JsonScalar = string | number | boolean | null;
export type JsonValue = JsonScalar | JsonValue[] | { [key: string]: JsonValue };
export type QueryParams = Record<string, JsonScalar | JsonScalar[]>;

const ALLOWED_HOSTNAMES = new Set(["dadosabertos.compras.gov.br", "pncp.gov.br"]);
const RETRYABLE_STATUSES = new Set([429, 502, 503, 504]);
const RETRYABLE_KINDS = new Set(["UPSTREAM_TIMEOUT", "UPSTREAM_UNAVAILABLE"]);
const RETRY_BASE_SECONDS = 0.2;
const RETRY_JITTER_RATIO = 0.1;
const MAX_ERROR_BODY_CHARS = 4096;

export class UpstreamError extends Error {
  readonly kind: string;
  readonly status: number | null;
  readonly provider: string | null;
  readonly endpoint: string | null;
  readonly upstreamMessage: string;
  readonly retryable: boolean;

  constructor(
    kind: string,
    status: number | null,
    message: string,
    opts?: { provider?: string; endpoint?: string; upstreamMessage?: string },
  ) {
    super(`${kind}: ${message}`);
    this.name = "UpstreamError";
    this.kind = kind;
    this.status = status;
    this.provider = opts?.provider ?? null;
    this.endpoint = opts?.endpoint ?? null;
    this.upstreamMessage = (opts?.upstreamMessage ?? message).slice(0, MAX_ERROR_BODY_CHARS);
    this.retryable =
      (status !== null && RETRYABLE_STATUSES.has(status)) || RETRYABLE_KINDS.has(kind);
  }
}

export interface ReadOnlyHttpOptions {
  maxRetries?: number;
  timeout?: number;
  maxDocumentBytes?: number;
  maxConcurrency?: number;
  requestsPerSecond?: number;
  provider?: string;
  fetchImpl?: typeof globalThis.fetch;
}

class Semaphore {
  private available: number;
  private readonly waiters: (() => void)[] = [];

  constructor(max: number) {
    this.available = max;
  }

  acquire(): Promise<void> {
    if (this.available > 0) {
      this.available -= 1;
      return Promise.resolve();
    }
    return new Promise<void>((resolve) => this.waiters.push(resolve));
  }

  release(): void {
    const next = this.waiters.shift();
    if (next !== undefined) next();
    else this.available += 1;
  }
}

interface RequestLimiter {
  semaphore: Semaphore | null;
  requestInterval: number | null;
  nextRequestAt: number;
  rateLock: Promise<void>;
}

// ponytail: single-threaded Node, one process-wide map keyed by origin beats Python's per-loop WeakKeyDictionary
const LIMITERS = new Map<string, RequestLimiter>();

function sleep(seconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, seconds * 1000));
}

export class ReadOnlyHttpClient {
  private readonly maxRetries: number;
  private readonly timeout: number;
  private readonly maxDocumentBytes: number;
  private readonly maxConcurrency: number | undefined;
  private readonly requestsPerSecond: number | undefined;
  private readonly base: string;
  private readonly origin: string;
  private readonly provider: string;
  private readonly fetchImpl: typeof globalThis.fetch;

  constructor(baseUrl: string, options: ReadOnlyHttpOptions = {}) {
    const maxRetries = options.maxRetries ?? 3;
    const maxDocumentBytes = options.maxDocumentBytes ?? 25_000_000;
    if (maxRetries < 0) {
      throw new Error("max_retries must be non-negative");
    }
    if (!Number.isInteger(maxDocumentBytes) || maxDocumentBytes <= 0) {
      throw new Error("max_document_bytes must be a positive finite integer");
    }
    if (
      options.maxConcurrency !== undefined &&
      (!Number.isInteger(options.maxConcurrency) || options.maxConcurrency <= 0)
    ) {
      throw new Error("max_concurrency must be positive");
    }
    if (
      options.requestsPerSecond !== undefined &&
      (!Number.isFinite(options.requestsPerSecond) || options.requestsPerSecond <= 0)
    ) {
      throw new Error("requests_per_second must be positive");
    }

    let parsed: URL;
    try {
      parsed = new URL(baseUrl);
    } catch {
      throw new Error("base_url must use a supported HTTPS port");
    }
    const hostname = parsed.hostname.toLowerCase();
    if (
      parsed.protocol !== "https:" ||
      !ALLOWED_HOSTNAMES.has(hostname) ||
      !(parsed.port === "" || parsed.port === "443") ||
      parsed.username !== "" ||
      parsed.password !== "" ||
      parsed.search !== "" ||
      parsed.hash !== ""
    ) {
      throw new Error("base_url must use an allowlisted HTTPS host and supported port");
    }

    this.maxRetries = maxRetries;
    this.timeout = options.timeout ?? 30;
    this.maxDocumentBytes = maxDocumentBytes;
    this.maxConcurrency = options.maxConcurrency;
    this.requestsPerSecond = options.requestsPerSecond;
    this.base = baseUrl.replace(/\/+$/, "");
    this.origin = `https://${hostname}`;
    this.provider = options.provider ?? (hostname === "pncp.gov.br" ? "pncp" : "compras");
    this.fetchImpl = options.fetchImpl ?? globalThis.fetch.bind(globalThis);
  }

  async get(path: string, params?: QueryParams): Promise<JsonValue | string> {
    const requestPath = ReadOnlyHttpClient.relativePath(path);
    const endpoint = `/${requestPath}`;
    const requestId = randomBytes(16).toString("hex");

    const url = new URL(`${this.base}/${requestPath}`);
    if (params !== undefined) {
      for (const [key, value] of Object.entries(params)) {
        const values = Array.isArray(value) ? value : [value];
        for (const item of values) {
          // ponytail: httpx renders None as bare key; URLSearchParams renders empty value
          url.searchParams.append(key, item === null ? "" : String(item));
        }
      }
    }

    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      let retry = false;
      let statusCode: number | null = null;
      const limiter = this.getLimiter();
      if (limiter?.semaphore != null) await limiter.semaphore.acquire();
      let started = Date.now();
      try {
        await this.waitForRateLimit(limiter);
        // parity com http_readonly.py:151: o timer mede apenas o pedido real, não a espera do rate limit
        started = Date.now();
        // ponytail: redirect:"manual" mirrors httpx not following redirects; allowlisted APIs answer 2xx/4xx directly
        const response = await this.fetchImpl(url.toString(), {
          method: "GET",
          redirect: "manual",
          signal: AbortSignal.timeout(Math.max(this.timeout * 1000, 1)),
        });
        statusCode = response.status;
        const body = await this.readBody(response, endpoint);
        if (RETRYABLE_STATUSES.has(response.status)) {
          if (attempt < this.maxRetries) {
            retry = true;
          } else {
            throw new UpstreamError(
              ReadOnlyHttpClient.errorKind(response.status),
              response.status,
              ReadOnlyHttpClient.errorBodyText(response, body),
              { provider: this.provider, endpoint },
            );
          }
        } else if (response.status === 404) {
          throw new UpstreamError("NOT_FOUND", 404, ReadOnlyHttpClient.errorBodyText(response, body), {
            provider: this.provider,
            endpoint,
          });
        } else if (response.status >= 300) {
          throw new UpstreamError(
            ReadOnlyHttpClient.errorKind(response.status),
            response.status,
            ReadOnlyHttpClient.errorBodyText(response, body),
            { provider: this.provider, endpoint },
          );
        }

        if (!retry) {
          if (ReadOnlyHttpClient.isJson(response)) {
            try {
              return JSON.parse(ReadOnlyHttpClient.bodyText(response, body)) as JsonValue;
            } catch (error) {
              throw new UpstreamError(
                "UPSTREAM_SCHEMA_CHANGED",
                response.status,
                error instanceof Error ? error.message : String(error),
                { provider: this.provider, endpoint },
              );
            }
          }
          return ReadOnlyHttpClient.bodyText(response, body);
        }
      } catch (error) {
        if (error instanceof UpstreamError) throw error;
        const message = error instanceof Error ? error.message : String(error);
        const isTimeout =
          error instanceof Error && (error.name === "TimeoutError" || error.name === "AbortError");
        if (attempt < this.maxRetries) {
          retry = true;
        } else {
          throw new UpstreamError(
            isTimeout ? "UPSTREAM_TIMEOUT" : "UPSTREAM_UNAVAILABLE",
            null,
            message,
            { provider: this.provider, endpoint },
          );
        }
      } finally {
        if (limiter?.semaphore != null) limiter.semaphore.release();
        logger.info("upstream request", {
          request_id: requestId,
          provider: this.provider,
          endpoint: createHash("sha256").update(endpoint, "utf8").digest("hex"),
          duration_ms: Math.round((Date.now() - started) * 1000) / 1000,
          status_code: statusCode,
          retry_count: attempt,
          cache_hit: false,
        });
      }
      if (retry) {
        await ReadOnlyHttpClient.sleepBeforeRetry(attempt);
        continue;
      }
    }

    throw new Error("unreachable");
  }

  private getLimiter(): RequestLimiter | null {
    let limiter = LIMITERS.get(this.origin);
    if (limiter === undefined) {
      if (this.maxConcurrency === undefined && this.requestsPerSecond === undefined) {
        return null;
      }
      limiter = {
        semaphore: this.maxConcurrency !== undefined ? new Semaphore(this.maxConcurrency) : null,
        requestInterval:
          this.requestsPerSecond !== undefined ? 1 / this.requestsPerSecond : null,
        nextRequestAt: 0,
        rateLock: Promise.resolve(),
      };
      LIMITERS.set(this.origin, limiter);
    }
    return limiter;
  }

  private async waitForRateLimit(limiter: RequestLimiter | null): Promise<void> {
    if (limiter === null || limiter.requestInterval === null) return;
    const previous = limiter.rateLock;
    let release!: () => void;
    limiter.rateLock = new Promise<void>((resolve) => {
      release = resolve;
    });
    await previous;
    try {
      const now = Date.now();
      if (limiter.nextRequestAt > now) {
        await sleep((limiter.nextRequestAt - now) / 1000);
      }
      limiter.nextRequestAt = Date.now() + limiter.requestInterval * 1000;
    } finally {
      release();
    }
  }

  private static relativePath(path: string): string {
    let decoded = path;
    for (;;) {
      let next: string;
      try {
        next = decodeURIComponent(decoded);
      } catch {
        break;
      }
      if (next === decoded) break;
      decoded = next;
    }
    let parsed: URL | null = null;
    try {
      parsed = new URL(`http://x${decoded}`);
    } catch {
      parsed = null;
    }
    if (
      !path.startsWith("/") ||
      parsed === null ||
      parsed.protocol !== "http:" ||
      parsed.host !== "x" ||
      decoded.startsWith("//") ||
      parsed.search !== "" ||
      parsed.hash !== "" ||
      decoded.includes("\\")
    ) {
      throw new Error("path must be relative to the fixed upstream origin");
    }
    // new URL() normalizes ".." away, so the segment check must run on the raw decoded path
    const decodedPath = decoded.split(/[?#]/)[0] ?? "";
    if (decodedPath.split("/").includes("..")) {
      throw new Error("path must be relative to the fixed upstream origin");
    }
    return path.replace(/^\/+/, "");
  }

  private static isJson(response: Response): boolean {
    const contentType = (response.headers.get("content-type") ?? "").split(";", 1)[0]
      ?.trim()
      .toLowerCase();
    return contentType === "application/json" || (contentType?.endsWith("+json") ?? false);
  }

  private static errorKind(status: number): string {
    if (status === 429) return "UPSTREAM_RATE_LIMIT";
    if (status >= 500) return "UPSTREAM_UNAVAILABLE";
    return "UPSTREAM_BAD_REQUEST";
  }

  private static async sleepBeforeRetry(attempt: number): Promise<void> {
    const delay = RETRY_BASE_SECONDS * 2 ** attempt;
    await sleep(delay + Math.random() * delay * RETRY_JITTER_RATIO);
  }

  private async readBody(response: Response, endpoint: string): Promise<Uint8Array> {
    const contentLength = response.headers.get("content-length");
    if (contentLength !== null) {
      const declared = Number(contentLength);
      if (Number.isInteger(declared) && declared > this.maxDocumentBytes) {
        throw new UpstreamError(
          "DOCUMENT_TOO_LARGE",
          response.status,
          `response exceeds ${this.maxDocumentBytes} bytes`,
          { provider: this.provider, endpoint },
        );
      }
    }

    if (response.body === null) return new Uint8Array(0);
    const reader = response.body.getReader();
    const chunks: Uint8Array[] = [];
    let total = 0;
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      total += value.byteLength;
      if (total > this.maxDocumentBytes) {
        throw new UpstreamError(
          "DOCUMENT_TOO_LARGE",
          response.status,
          `response exceeds ${this.maxDocumentBytes} bytes`,
          { provider: this.provider, endpoint },
        );
      }
      chunks.push(value);
    }
    const body = new Uint8Array(total);
    let offset = 0;
    for (const chunk of chunks) {
      body.set(chunk, offset);
      offset += chunk.byteLength;
    }
    return body;
  }

  private static bodyText(response: Response, body: Uint8Array): string {
    const contentType = response.headers.get("content-type") ?? "";
    const match = /(?:^|;)\s*charset\s*=\s*"?([^";]+)"?/i.exec(contentType);
    const charset = (match?.[1] ?? "utf-8").trim();
    const decode = (label: string): string => new TextDecoder(label, { fatal: false }).decode(body);
    try {
      return decode(charset === "" ? "utf-8" : charset);
    } catch {
      return decode("utf-8");
    }
  }

  private static errorBodyText(response: Response, body: Uint8Array): string {
    return ReadOnlyHttpClient.bodyText(response, body).slice(0, MAX_ERROR_BODY_CHARS);
  }
}
